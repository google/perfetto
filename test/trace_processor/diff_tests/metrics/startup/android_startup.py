#!/usr/bin/env python3
# Copyright (C) 2018 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from os import sys, path

import synth_common

trace = synth_common.create_trace()
trace.add_packet()
trace.add_process(1, 0, 'init')
trace.add_process(2, 1, 'system_server')
trace.add_process(3, 1, 'com.google.android.calendar', 10001)
trace.add_process(4, 3, 'com.google.android.calendar', 10001)

trace.add_package_list(
    ts=1, name='com.google.android.calendar', uid=10001, version_code=123)

trace.add_ftrace_packet(cpu=0)
# Intent without any corresponding end state, will be ignored
trace.add_atrace_begin(
    ts=100, tid=2, pid=2, buf='MetricsLogger:launchObserverNotifyIntentStarted')
trace.add_atrace_end(ts=101, tid=2, pid=2)

# Start intent for a successful launch of calendar
trace.add_atrace_begin(
    ts=102, tid=2, pid=2, buf='MetricsLogger:launchObserverNotifyIntentStarted')
trace.add_atrace_end(ts=103, tid=2, pid=2)

trace.add_atrace_async_begin(
    ts=110, tid=2, pid=2, buf='launching: com.google.android.calendar')

trace.add_sched(ts=110, prev_pid=0, next_pid=3)

# As the process already existed before intent started, this is a
# warm/hot start (we choose warm). Therefore, emit an activityStart
# slice.
trace.add_atrace_begin(ts=115, tid=3, pid=3, buf='activityStart')
trace.add_atrace_end(ts=117, tid=3, pid=3)
trace.add_atrace_begin(ts=117, tid=3, pid=3, buf='activityResume')
trace.add_atrace_end(ts=118, tid=3, pid=3)

# P1: 10ns running
trace.add_sched(ts=120, prev_pid=3, next_pid=0, prev_state='S')
# P1: 10ns sleep
trace.add_sched(ts=130, prev_pid=0, next_pid=3)

trace.add_sched(ts=130, prev_pid=3, next_pid=4)

# Create an unrelated task
trace.add_newtask(ts=155, tid=1, new_tid=5, new_comm='', flags=0)

# P2: 30ns running
trace.add_sched(ts=160, prev_pid=4, next_pid=0, prev_state='R')
# P2: 49ns runnable
trace.add_sched(ts=209, prev_pid=0, next_pid=4)
# P2: 1ns running
trace.add_sched(ts=210, prev_pid=4, next_pid=0)

trace.add_atrace_async_end(
    ts=210, tid=2, pid=2, buf='launching: com.google.android.calendar')
trace.add_atrace_begin(
    ts=211,
    tid=2,
    pid=2,
    buf='MetricsLogger:launchObserverNotifyActivityLaunchFinished')
trace.add_atrace_end(ts=212, tid=2, pid=2)

# Some time after, add a slice for fully drawn frame.
trace.add_atrace_begin(
    ts=300,
    tid=3,
    pid=3,
    buf='reportFullyDrawn() for {{com.google.android.calendar}}')
trace.add_atrace_end(ts=305, tid=2, pid=2)

# Start intent for calendar, we failed to launch the activity.
trace.add_atrace_begin(
    ts=402, tid=2, pid=2, buf='MetricsLogger:launchObserverNotifyIntentStarted')
trace.add_atrace_end(ts=403, tid=2, pid=2)

trace.add_atrace_async_begin(
    ts=410, tid=2, pid=2, buf='launching: com.google.android.calendar')

trace.add_atrace_async_end(
    ts=510,
    tid=2,
    pid=2,
    buf='launching: com.google.android.apps.nexuslauncher')

trace.add_ftrace_packet(cpu=1)
trace.add_sched(ts=160, prev_pid=0, next_pid=1)
trace.add_sched(ts=200, prev_pid=1, next_pid=0)

sys.stdout.buffer.write(trace.trace.SerializeToString())
