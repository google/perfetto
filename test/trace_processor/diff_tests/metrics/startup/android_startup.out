android_startup {
  startup {
    startup_id: 0
    cpu_count: 2
    package_name: "com.google.android.calendar"
    process_name: "com.google.android.calendar"
    zygote_new_process: false
    to_first_frame {
      dur_ns: 108
      main_thread_by_task_state {
        running_dur_ns: 10
        runnable_dur_ns: 80
        uninterruptible_sleep_dur_ns: 0
        interruptible_sleep_dur_ns: 10
        uninterruptible_io_sleep_dur_ns: 0
        uninterruptible_non_io_sleep_dur_ns: 0
      }
      other_processes_spawned_count: 1
      time_activity_manager {
        dur_ns: 8
        dur_ms: 8e-06
      }
      time_activity_start {
        dur_ns: 2
        dur_ms: 2e-06
      }
      time_activity_resume {
        dur_ns: 1
        dur_ms: 1e-06
      }
      dur_ms: 0.000108
      time_to_running_state {
        dur_ns: 8
        dur_ms: 8e-06
      }
    }
    activity_hosting_process_count: 1
    process {
      name: "com.google.android.calendar"
      uid: 10001
      package {
        package_name: "com.google.android.calendar"
        apk_version_code: 123
        debuggable: false
      }
      pid: 3
      android_user_id: 0
      is_kernel_task: false
    }
    report_fully_drawn {
      dur_ns: 198
      dur_ms: 0.000198
    }
    event_timestamps {
      intent_received: 102
      first_frame: 210
    }
    system_state {
      dex2oat_running: false
      installd_running: false
      broadcast_dispatched_count: 0
      broadcast_received_count: 0
      most_active_non_launch_processes: "init"
      most_active_non_launch_processes: "com.google.android.calendar"
      installd_dur_ns: 0
      dex2oat_dur_ns: 0
    }
    slow_start_reason_with_details {
      reason_id: MAIN_THREAD_TIME_SPENT_IN_RUNNABLE
      reason: "Main Thread - Time spent in Runnable state"
      severity: WARNING
      expected_value {
        value: 15
        unit: PERCENTAGE
        higher_expected: false
      }
      actual_value {
        value: 74
        dur: 80
      }
      launch_dur: 108
      trace_thread_sections {
        thread_section {
          start_timestamp: 130
          end_timestamp: 210
          thread_name: "com.google.android.calendar"
          process_pid: 3
          thread_tid: 3
        }
        start_timestamp: 130
        end_timestamp: 210
      }
    }
    startup_type: "warm"
  }
}
